torch/serialization.py

```diff
@@ -291,6 +291,25 @@ def raise_err_msg(patterns, e):
     except (io.UnsupportedOperation, AttributeError) as e:
         raise_err_msg(["seek", "tell"], e)
 
+def _check_dill_version(pickle_module):
+    '''Checks if using dill as the pickle module, and if so, checks if it is the correct version.
+    If dill version is lower than 0.3.1, a ValueError is raised.
+
+    Args:
+        pickle_module: module used for pickling metadata and objects
+
+    '''
+    if pickle_module.__name__ == 'dill':
+        required_dill_version = (0, 3, 1)
+        dill_version = tuple(int(num) for num in pickle_module.__version__.split('.'))
```

Comment:
We're assuming that the dill version is always well defined, like "0.3.1". PyTorch master versions don't look as nice, how common is it for python package version strings to not be well defined?

Reply:
It's uncommon for packages to not have `'x.y.z'`, however the Python version spec (PEP 440) does allow for development version numbers like `'0.3.1.dev0+3141324'` which will fail here with
```
ValueError: invalid literal for int() with base 10: 'dev0+3141324'
```
Similarly, there's `.postX` qualifiers (less often used), or `rc1` or `b1` (release candidate, beta) post-fixes to the `z` part of the `x, y, z` version string. And I see that the current version of `dill` is `'0.3.1.1'`, which works with this PR but is not PEP 440 compliant.

In general comparing Python versions is a real pain to get 100% right. Also `setuptools`'s `LooseVersion` and `StrictVersion` are both broken in subtle ways. For numpy I implemented a `NumpyVersion` class that should work correctly ([here]()https://github.com/numpy/numpy/blob/master/numpy/lib/_version.py#L18), but that's ~100 LoC and probably overkill for this.

My suggestion would be to be pragmatic and just add a minor tweak to what you have here. Perhaps:
```
dill_version = pickle_module.__version__.split('.')
# since the `Z` in a Python 'X.Y.Z' can be messed up in various ways and contain
# 'dev', 'post', '+' or other characters, try first: 
tuple(int(num) for num in dill_version[:2])
# Then check if first character of `Z` is a number. If so, check it's >= 1. 
# If it's `'0'` or not a number, raise a ValueError
```

Reply:
If we can't parse the dill version, it might be better to throw a warning instead of raising an error. It would be bad for users to have to build pytorch from source (well, they just need to modify the python file inside the installation...) if the dill version was malformed

Reply:
What do you guys think about something like this? This function tries to pull out the same number of fields that were provided in the required version tuple, then convert each field to match the type of the corresponding requirement field, and then finally perform a comparison of the two tuples whose types and size match each other.

If the module version string has fewer fields, or if they cannot be converted to match the requirement types, then an exception would be raised.

```
    def check_module_version_greater_or_equal(module, req_version_tuple):
        module_version_strs = module.__version__.split('.')[:len(req_version_tuple)]

        # Cast module version fields to match the types of the required version
        module_version = tuple(
            type(req_field)(module_version_strs[idx]) for idx, req_field in enumerate(req_version_tuple)
        )
        return module_version >= req_version_tuple
```
The caller of this function can wrap it in a try/except, and do something differently depending on what needs to be done if the version string is malformed.

For instance, a unit test should probably fail with an error if the string was malformed, since we could not determine if the dill version is high enough. If the caller is torch.load/save, we can emit a warning, but then still attempt to perform the load/save since we don't know for sure if the dill module will fail to do so.

Does that seem like a suitable approach?

Reply:
@kurtamohler I don't have a preference between your proposed code, the original code in this PR, and rgommers's proposal. What is important to me is that there is some graceful handling if the parsing fails